### PR TITLE
Advanced search enhance

### LINKED ISSUE #236: Java SDK - Advanced Search Enhancement
Working on the Advanced Search Feature for Java SDK, will be submitting a pull request today . May 6 2016 with my changes.

If this is the wrong type of request feel free to change

### PR DESCRIPTION
**Pull request by rschaller-box**
*Note Implemented Recommended Changes from comments found : *
#237]

Add search/advancedSearch

Introduced a new BoxSearch class that uses the /search API endpoint. It also allow Advanced Search Capabilities.

Closes #236

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @rschaller-box has signed the CLA. Thanks for the pull request!

File: src/main/java/com/box/sdk/BoxSearchParameters.java [hunk 1]
```diff
@@ -318,7 +317,7 @@ private String printParam(Object obj)    {
      * @param    paramValue Object that will be checked if null.
      * @return this.true if the parameter that is being checked is not null
      */
-    public boolean isNotNull(Object paramValue) {
+    private boolean isNullOrEmpty(Object paramValue) {
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxSearchParameters.java]: The logic of this method needs to be flipped. Right now it returns true if the object is not null.

File: src/main/java/com/box/sdk/BoxSearchParameters.java [hunk 2]
```diff
@@ -288,15 +291,11 @@ public void setTrashContent(String trashContent) {
...
+     * Retrieve the existing BoxMetaDataFilter.
+     * @return this.BoxMetaDataFilter
      */
-    public BoxMetadataFilterRequest getMdFilters() {
-        BoxMetadataFilterRequest bmfr = new BoxMetadataFilterRequest();
-        for (BoxMetadataFilter bmf : this.mdFilters) {
-            bmfr.addBoxMetadataFilter(bmf);
-        }
-        return bmfr;
+    public List<BoxMetadataFilter> getMdFilters() {
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxSearchParameters.java]: I think we should rename this to `getMetadataFilters()` to make the name clearer.

File: src/main/java/com/box/sdk/BoxMetadataFilter.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+    public JsonObject getFiltersList() {
+        return this.filtersList;
+    }
+    /**
+     * Set a filter to the filterList, example: key=documentType, value=special.
+     * @param key the key that the filter should be looking for.
+     * @param value the specific value that corresponds to the key.
+     */
+    public void setFilter(String key, String value) {
+        this.filtersList = new JsonObject();
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxMetadataFilter.java]: Can this class can only have one filter type set at a time? It looks like `filtersList` is being recreated in each setter.

File: src/main/java/com/box/sdk/BoxSearch.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    /**
+     * Searches all descendant folders using a given query and query parameters.
+     * @param  offset is the starting position.
+     * @param  limit the number of search results CANNONT Exceed 1000.
+     * @param  bsp containing query and advanced search capabilities.
+     * @return a PartialCollection containing the search results.
+     */
+    public PartialCollection<BoxItem.Info> search(long offset, long limit, final BoxSearchParameters bsp) {
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxSearch.java]: Rename to `searchRange` to be more consistent with `BoxFolder`. That way we can add another search method later which can do automatic paging.

File: src/main/java/com/box/sdk/BoxSearchParameters.java [hunk 5]
```diff
@@ -0,0 +1,442 @@
...
+     * @param fileExtensions applied as a filter for extensions.
+     */
+    public void setFileExtensions(List<String> fileExtensions) {
+        this.fileExtensions = fileExtensions;
+    }
+    /**
+     * Get the from date filter to specify the when a file was created.
+     * @return this.created date from.
+     */
+    public Date getCreatedAtRangeFromDate() {
```
Comment (gcurtis) [inline src/main/java/com/box/sdk/BoxSearchParameters.java]: The names of the range methods are kind of confusing. If you create `DateRange` and `SizeRange` classes, then it'll be a bit clearer.

Usage example:

```java
DateRange range = new DateRange(from, to);
params.setCreatedAtRange(range);
```

**Comment by rschaller-box** [comment]
@gcurtis please review I believe we are good to go :)

**Comment by amjoesph** [comment]
Does this advanced search functionality include these features:
- pagination of records 
- retrieve list of folders/files only on a given folder id

Does the search result would provide the path of a file.

**Comment by gcurtis** [comment]
@amjoesph it doesn't automatically handle paging results yet, but it will still support retrieving items from a specific folder ID.

**Comment by gcurtis** [comment]
Squashed and merged in 11e97cce2ecd56c8dfe81a0a3c1c2fa22317d64a.

**Pull request closed without merging**